Improve guidance for testing vision service (#1616)
viamrobotics/docs: docs/services/vision/detection.md

{{% /alert %}}

Click **Save config** and head to the **Components** tab.

> Proceed to [Add a camera component and a "transform" model](#test-your-detector).

Reviewer: ```suggestion
Proceed to [Test your detector](#test-your-detector).
```

{{% /tab %}}
{{< /tabs >}}

Click **Save config**.

## Test your detector

You can test your detector with [live camera footage](#live-camera-footage) or [existing images](#existing-images).

> ### Live Camera footage

Reviewer: ```suggestion
### Live camera footage
```

If you intend to use the detector with a camera that is part of your robot, you can test your detector from the [**Control tab**](/manage/fleet/robots/#control) or with code:

1. Configure a physical [camera component](../../../components/camera/).
2. If you would like to see detections from the **Control tab**, configure a [transform camera](../../../components/camera/transform/) to view output from the detector overlaid on images from the physical camera.
   If you only want to access detections from code, you do not need to configure a transform camera.
3. After adding the components and their attributes, click **Save config**.
4. Navigate to the **Control** tab, click on your transform camera and toggle it on.
   The transform camera will now show detections with bounding boxes around the object.

> ![Viam app control tab interface showing bounding boxes around two office chairs, both labeled "chair" with confidence score "0.50."](/services/vision/chair-detector.png)

Author: ```suggestion
   ![Viam app control tab interface showing bounding boxes around two office chairs, both labeled "chair" with confidence score "0.50."](/services/vision/chair-detector.png)
```

    Wait for the robot to reload, and then go to the **Control** tab and open the transform camera card to test the stream of detections.
1. Configure a physical [camera component](../../../components/camera/).
2. If you would like to see detections from the **Control tab**, configure a [transform camera](../../../components/camera/transform/) to view output from the detector overlaid on images from the physical camera.
   If you only want to access detections from code, you do not need to configure a transform camera.
3. After adding the components and their attributes, click **Save config**.
4. Navigate to the **Control** tab, click on your transform camera and toggle it on.
   The transform camera will now show detections with bounding boxes around the object.

![Viam app control tab interface showing bounding boxes around two office chairs, both labeled "chair" with confidence score "0.50."](/services/vision/chair-detector.png)

## Code

The following code gets the robot’s vision service and then runs a color detector vision model on output from the robot's camera `"cam1"`.

> {{% alert title="Tip" color="tip" %}}

Reviewer: Why get rid of this alert? I added it because of the user feedback (via Fahmina) that this is a common mistake

Reviewer: Perhaps my line 235 suggestion is a way to address this.

Author: Because it doesn't have to be a physical camera - you could have a transform camera that crops the image and then pass that. So specifically calling out that it is not the name of the transform camera is weird.

Reviewer 2: I think the thing that needs to be emphasized is that when setting up the vision service on python, you shouldn't call "get_classifications" or "get_detections" using the camera that already has the "detections" or "classifications" transform applied to it -- it would be like calling "Detections" on a camera that already called "Detections"

Reviewer 2: So yea, if your transform camera just is doing a crop, then using get_detections_from_camera on it is fine, but I do remember that people have made a mistake of sometimes creating a transform camera that applies detections, and then in python, using that camera to call get_detections again (which creates a weird output and a slow camera)

Author: Ok - I've aded that as an alert - I like that way of putting it

> The following code gets the robot’s vision service and then runs a color detector vision model on output from the robot's camera `"cam1"`:

Reviewer: ```suggestion
5. To access detections with code, use the Vision Service methods on the physical camera you configured (not the transform camera).
The following code gets the robot’s vision service and then runs a color detector vision model on output from the robot's camera `"cam1"`:
```
I think it's a bit odd because we have three options (live stream control tab, live stream code, and preexisting images with code), but  they're sort of lumped into 2 instead. I think making this a step may help it make more sense.

{{< /tabs >}}

Click **Save config**.

## Test your detector

You can test your detector with [live camera footage](#live-camera-footage) or [existing images](#existing-images).

### Live camera footage

If you intend to use the detector with a camera that is part of your robot, you can test your detector from the [**Control tab**](/manage/fleet/robots/#control) or with code:

1. Configure a physical [camera component](../../../components/camera/).
2. If you would like to see detections from the **Control tab**, configure a [transform camera](../../../components/camera/transform/) to view output from the detector overlaid on images from the physical camera.
   If you only want to access detections from code, you do not need to configure a transform camera.
3. After adding the components and their attributes, click **Save config**.
4. Navigate to the **Control** tab, click on your transform camera and toggle it on.
   The transform camera will now show detections with bounding boxes around the object.

   ![Viam app control tab interface showing bounding boxes around two office chairs, both labeled "chair" with confidence score "0.50."](/services/vision/chair-detector.png)

> 5. To access detections with code, use the Vision Service methods on the physical camera you configured (not the transform camera).

Author: ```suggestion
5. To access detections with code, use the Vision Service methods on the camera you configured.
Do not use the name of the transform camera that you added in step 2.
```

    ```json
    {
      "pipeline": [
          {
          "type": "detections",
          "attributes": {
              "confidence_threshold": 0.5,
              "detector_name": "my_detector"
          }
          }
      ],
      "source": "<camera-name>"
    }
    ```

3. After adding the components and their attributes, click **Save config**.
4. Navigate to the **Control** tab, click on your transform camera and toggle it on.
   The transform camera will now show detections with bounding boxes around the object.

   ![Viam app control tab interface showing bounding boxes around two office chairs, both labeled "chair" with confidence score "0.50."](/services/vision/chair-detector.png)

5. To access detections with code, use the Vision Service methods on the camera you configured in step 1.
   The following code gets the robot’s vision service and then runs a color detector vision model on output from the robot's camera `"cam1"`:

   {{% alert title="Tip" color="tip" %}}

   Pass the name of the camera you configured in step 1.

>    Do not pass a transform camera that already already has the "detections" or "classifications" transform applied to it.

Author: ```suggestion
   Do not pass a transform camera that already has the "detections" or "classifications" transform applied to it.
```

detections = await my_detector.get_detections_from_camera(cam1)

> # If you need to use the image based on the detections, get the image first

Reviewer: ```suggestion
# If you need to save the image, get the image first
```
[opt] Not sure how to reword (I don't love my suggestion), but I think it's kinda unclear this way

Reviewer: What I mean here is that maybe you just need to use the image period, regardless of the detections.

detections = await my_detector.get_detections_from_camera(cam1)

> # If you need to save the image, get the image first
> # and then run detections on it. This process is slower:

Author: ```suggestion
# If you need to store the image, get the image first
# and then run detections on it. This process is slower:
```

> # and then run detections on it. This process is slower.

Reviewer: ```suggestion
# and then run detections on it. This process is slower:
```
I think the colon helps make clear that we are talking about the process below, as opposed to a process not shown that the user would have to write themselves.

    logger.Info(detections[0])
}

> // If you need to use the image based on the detections, get the image first

Reviewer: same as above [opt]

    logger.Info(detections[0])
}

> # If you need to store the image, get the image first
> # and then run detections on it. This process is slower:

Author: ```suggestion
// If you need to store the image, get the image first
// and then run detections on it. This process is slower:
```

> // and then run detections on it. This process is slower.

Reviewer: ```suggestion
// and then run detections on it. This process is slower:
```
Same as above

> // If you need to use the image based on the detections, get the image first
> // and then run detections on it. This process is slower:

Author: ```suggestion

# If you need to store the image, get the image first
# and then run detections on it. This process is slower:
```

# Load an image
img = Image.open('test-image.png')

# Apply the detector to the image
detections_from_image = await my_detector.get_detections(img)

await robot.close()
```

To learn more about how to use detection, see the [Python SDK docs](https://python.viam.dev/autoapi/viam/services/vision/index.html).

{{% /tab %}}
{{% tab name="Go" %}}

```go {class="line-numbers linkable-line-numbers"}
import (
  "go.viam.com/rdk/config"
  "go.viam.com/rdk/services/vision"
  "image/jpeg"
  "os"
)

visService, err := vision.from_robot(robot=robot, name='my_detector')
if err != nil {
    logger.Fatalf("Cannot get Vision Service: %v", err)
}

// Read image from existing file
file, err := os.Open("test-image.jpeg")
if err != nil {
    logger.Fatalf("Could not get image: %v", err)
}
defer file.Close()
img, err := jpeg.Decode(file)
if err != nil {
    logger.Fatalf("Could not decode image: %v", err)
}
defer img.Close()

// Apply the detector to the image
detectionsFromImage, err := visService.Detections(context.Background(), img, nil)
if err != nil {
    logger.Fatalf("Could not get detections: %v", err)
}
if len(detectionsFromImage) > 0 {
    logger.Info(detectionsFromImage[0])

> }

Author: I was not able to test this code - I get a session expiry error on all detections code currently

Author: nvm - tested and works
